Internal error in Immediate Window when trying to create an already e…

ivanbasov opened the pull request:
### Customer scenario

Customer debugs a code that has a variable declaration like
```
int x = 0;
System.Diagnostics.Debugger.Break();
```
At a breakpoint, customer tries to declare the same variable once again in the immediate window or in a watch window by executing
`int x = 1;`

Expected:
error CS0136: A local or parameter named 'x' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter

Actual:
Internal error in the C# compiler

### Bugs this fixes
https://github.com/dotnet/roslyn/issues/22599
https://github.com/dotnet/roslyn/issues/22474
#15132

### Workarounds, if any

None

### Risk

Low

### Performance impact

None

### Is this a regression from a previous update?
No

### How was the bug found?
Customer reported

AlekseyTs commented on src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EELocalSymbol.cs:
```diff
@@ -79,7 +79,7 @@ public override string Name
         internal override SyntaxToken IdentifierToken
         {
-            get { throw ExceptionUtilities.Unreachable; }
+            get { return default; }
```
>get { return default; } [](start = 12, length = 23)

Given the changes in ```ExtendBinderChain```, do we still need this change? Let's revert it otherwise.

ivanbasov replied:
Great! We do not need this change anymore. Reverting. Thanks!

---
In reply to: [161291140](https://github.com/dotnet/roslyn/pull/24198#discussion_r161291140) [](ancestors = 161291140)

AlekseyTs commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/DeclarationTests.cs:
```diff
@@ -2139,5 +2144,31 @@ .maxstack  4
         }
+
+        [Fact]
+        public void DuplicateDeclaration()
```
>DuplicateDeclaration [](start = 20, length = 20)

Is it possible to add similar test for VB?

ivanbasov replied:
Good question!
It looks that:
1. It does not allow to execute
'Dim x = 3'
in the immediate window even for non declared variable.
It returns:
Statements of type 'FieldDeclaration' are not allowed in the Immediate window.

2. If one tries to execute 'x = 3' in the immediate window, it would either use an existing 'x' from code or from previous statements executed, or introduce a new variable.

Therefore, it seems that there is no such problem in VB. So, I do not see a suitable test for VB.

---
In reply to: [161291906](https://github.com/dotnet/roslyn/pull/24198#discussion_r161291906) [](ancestors = 161291906)

AlekseyTs commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/DeclarationTests.cs:
```diff
@@ -2139,5 +2144,31 @@ .maxstack  4
         }
+
+        [Fact]
+        public void DuplicateDeclaration()
```
>DuplicateDeclaration [](start = 20, length = 20)

Please add similar test for ```out var``` scenario described in #22474.

AlekseyTs commented:
Done with review pass (iteration 1)

ivanbasov commented:
Thank you! Addressed feedback for iteration 1.

---
In reply to: [357315314](https://github.com/dotnet/roslyn/pull/24198#issuecomment-357315314) [](ancestors = 357315314)

cston commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/DeclarationTests.cs:
```diff
@@ -2139,5 +2144,59 @@ .maxstack  4
+
+            WithRuntimeInstance(compilation0, runtime =>
+            {
+                var context = CreateMethodContext(runtime, "C.M", atLineNumber: 999);
```
`#line 999` shouldn't be necessary for either this test or the previous test. In both cases, the existing locals are in scope.

ivanbasov commented:
Retargetted against Dev15.6.x

Tagging @MeiChin-Tsai (Compilers) and @Pilchie (Expression Evaluator) for Ask Mode Approval

Pilchie commented:
Approved on my side.  @MeiChin-Tsai ?

AlekseyTs commented:
@ivanbasov You might want to ping @MeiChin-Tsai via e-mail.

MeiChin-Tsai commented:
approved. thx.
